flutter-idea/src/io/flutter/run/daemon/DevToolsServerTask.java

```diff
@@ -0,0 +1,399 @@
+
+      // This is only for development to check integration with a locally run DevTools server.
+      // To enable, follow the instructions in:
+      // https://github.com/flutter/flutter-intellij/blob/master/CONTRIBUTING.md#developing-with-local-devtools
+      final String localDevToolsDir = Registry.stringValue(LOCAL_DEVTOOLS_DIR);
+      if (!localDevToolsDir.isEmpty()) {
+        progressIndicator.setFraction(60);
+        progressIndicator.setText2("Starting local server");
+        setUpLocalServer(localDevToolsDir);
+      }
+
+      // If the Dart plugin does not start DevTools, then call `dart devtools` to start the server.
+      final Boolean dartPluginStartsDevTools = true;
+      if (!dartPluginStartsDevTools) {
+        progressIndicator.setFraction(60);
+        progressIndicator.setText2("Starting server");
+        setUpWithDart(createCommand(DartSdk.getDartSdk(project).getHomePath(),
+                                    DartSdk.getDartSdk(project).getHomePath() + File.separatorChar + "bin" + File.separatorChar + "dart",
+                                    ImmutableList.of("devtools", "--machine")));
+      }
+
+      // Otherwise, wait for the Dart Plugin to start the DevTools server.
+      final CompletableFuture<DevToolsInstance> devToolsFuture = checkForDartPluginInitiatedDevToolsWithRetries(progressIndicator);
+      devToolsFuture.whenComplete((devTools, error) -> {
+        if (error != null) {
+          cancelWithError(new Exception(error));
+        }
+        else {
+          devToolsFutureRef.get().complete(devTools);
+        }
+      });
+    }
+    catch (java.util.concurrent.ExecutionException | InterruptedException e) {
+      cancelWithError(e);
+    }
+  }
+
+  private Boolean dartSdkSupportsDartDevTools() {
+    final DartSdk dartSdk = DartSdk.getDartSdk(project);
+    if (dartSdk != null) {
+      final Version version = Version.parseVersion(dartSdk.getVersion());
+      assert version != null;
+      return version.compareTo(2, 15, 0) >= 0;
+    }
+    return false;
+  }
+
+  @Override
+  public void onThrowable(@NotNull Throwable error) {
+    super.onThrowable(error);
+    cancelWithError(new Exception(error));
+  }
+
+  private void setUpLocalServer(String localDevToolsDir) throws java.util.concurrent.ExecutionException, InterruptedException {
+    final DtdUtils dtdUtils = new DtdUtils();
+    final DartToolingDaemonService dtdService = dtdUtils.readyDtdService(project).get();
+    final String dtdUri = dtdService.getUri();
+
+    final List<String> args = new ArrayList<>();
+    args.add("serve");
+    args.add("--machine");
+    args.add("--dtd-uri=" + dtdUri);
+    final String localDevToolsArgs = Registry.stringValue(LOCAL_DEVTOOLS_ARGS);
+    if (!localDevToolsArgs.isEmpty()) {
+      args.addAll(Arrays.stream(localDevToolsArgs.split(" ")).toList());
+    }
+
+    setUpInDevMode(createCommand(localDevToolsDir, "dt", args));
+  }
+
+  private void setUpInDevMode(GeneralCommandLine command) {
```

Comment:
Can we annotate this arg `@NotNull` and mark the return of `createCommand` as `@NotNull`?

Reply:
Yes, done